README
NAME

    cpangitify - Convert cpan distribution from BackPAN to a git repository

VERSION

    version 0.11

SYNOPSIS

     % cpangitify Foo::Bar

DESCRIPTION

    This script fetches all known revisions of a distribution from
    CPAN/BackPAN and creates a git repository with one revision and one tag
    for each version of the distribution.

    The idea is to create a starting point for a git work flow when
    adopting a CPAN module for which you don't have access to the original
    repository. It is of course better to import from Subversion or to
    clone an existing git repository, but that may not be an option.

    If the distribution you are migrating changed names during its history,
    simply specify each name it had on the command line. Be sure to specify
    the current name first (this will be used when creating a directory
    name for the repository). For example Mojolicious::Plugin::TtRenderer
    was once called MojoX::Renderer::TT, so you would get both names in the
    history like this:

     % cpangitify Mojolicious::Plugin::TtRenderer MojoX::Renderer::TT

OPTIONS

 --help | -h

    Print out help and exit.

 --version

    Print out version and exit.

CAVEATS

    Each commit belongs to the CPAN author who submitted the corresponding
    release, therefore git blame may not be that useful for the imported
    portion of your new repository history.

    The commits are ordered by date, so where there are interleaving of
    releases that belong to development and production branches this simple
    minded script will probably do the wrong thing.

    Patches are welcome.

SEE ALSO

    Here are some similar projects:

      * Git::CPAN::Patch

      Comes with a git cpan import which does something similar. With this
      incantation I was able to get a repository for YAML (including
      history, but without authors and without the correct dates):

       % mkdir YAML
       % git init .
       % git cpan import --backpan YAML
       % git merge remotes/cpan/master

      One advantage here over cpangitify is that you should then later be
      able to import/merge future CPAN releases into yours.
      Git::CPAN::Patch also has a bunch of other useful tools for creating
      and submitting patches and may be worth checking out.

      If you do an internet search for this sort of thing you may see
      references to git-backpan-init, but this does not appear to be part
      of the Git::CPAN::Patch anymore (I believe git-import with the
      --backpan option is the equivalent).

      In general cpangitify is a one trick poney (though good at that one
      thing), and Git::CPAN::Patch is a Batman's utility belt with
      documentation that (for me at least) is pretty impenetrable.

      * gitpan <https://github.com/gitpan>

      Doesn't appear to have been updated in a number of years.

      * ggoosen's cpan2git <https://github.com/ggoossen/cpan2git>

    The reason I am not using the latter two is that they are designed to
    mirror the whole of CPAN/BackPAN, but I'm mostly just interested in one
    or two distributions here and there.

AUTHOR

    Graham Ollis <plicease@cpan.org>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2013 by Graham Ollis.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for App-cpangitify

0.11      2014-10-10 11:41:51 -0400
  - 'git rm' before 'git add'
    This seems to help on MSWin32, where I may have a newer version of git than on Debian Linux

0.10      2014-08-05 13:22:41 -0400
  - use HTTP::Tiny instead of LWP
    the former being part of the core as of Perl 5.14
  - use JSON::PP instead of JSON
    the former being part of the core as of Perl 5.14
  - use CPAN::ReleaseHistory instead of BackPAN::Index
    the latter may (?) be faster,
    but the former has lighter set of prereqs

0.09      2014-06-19 16:05:18 -0400
  - CPAN release (0.08 never made it to CPAN)

0.08      2014-01-08 06:10:17 -0500
  - make multiple tags a warning only
  - install with Module::Build instead of MakeMaker
  - use Archive::Extract instead of tar
  - use recursive Path::Class::File#remove and Path::Class::Dir#remove instead of calling system rm -rf

0.07      2013-10-18 12:41:40 -0400
  - make git 1.5.0 or better a requirement

0.06      2013-10-13 15:12:49 -0400
  - documentation (mention Git::CPAN::Patch)

0.05      2013-10-13 05:39:34 -0400
  - require BackPAN::Index 0.41

0.04_01   2013-10-11 17:38:22 -0400
  - include git version in test diagnostic

0.04      2013-10-09 20:29:45 -0400
  - change name to cpangitify
  - handle distros with name changes

0.03      2013-10-09 15:43:54 -0400
  - documentation

0.02      2013-10-09 13:57:30 -0400
  - include date and author info from cpan

0.01      2013-10-09 12:58:19 -0400
  - initial version


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
use strict;
use warnings;

use Module::Build 0.28;
use lib qw{inc}; use My::ModuleBuild;

my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "Git::Wrapper" => 0,
    "Module::Build" => "0.28",
    "Sort::Versions" => 0
  },
  "dist_abstract" => "Convert cpan distribution from BackPAN to a git repository",
  "dist_author" => [
    "Graham Ollis <plicease\@cpan.org>"
  ],
  "dist_name" => "App-cpangitify",
  "dist_version" => "0.11",
  "license" => "perl",
  "module_name" => "App::cpangitify",
  "recommends" => {
    "PerlX::Maybe::XS" => 0
  },
  "recursive_test_files" => 1,
  "requires" => {
    "Archive::Extract" => 0,
    "CPAN::ReleaseHistory" => 0,
    "File::Copy::Recursive" => 0,
    "File::chdir" => 0,
    "Git::Wrapper" => 0,
    "HTTP::Tiny" => 0,
    "IPC::System::Simple" => 0,
    "JSON::PP" => 0,
    "Path::Class" => "0.26",
    "PerlX::Maybe" => "0.003",
    "URI" => 0,
    "autodie" => 0,
    "perl" => "v5.10.0"
  },
  "script_files" => [
    "bin/cpangitify"
  ],
  "test_requires" => {
    "Capture::Tiny" => 0,
    "File::HomeDir" => "0.91",
    "HTTP::Date" => 0,
    "Test::More" => "0.94",
    "URI::file" => 0
  }
);


my %fallback_build_requires = (
  "Capture::Tiny" => 0,
  "File::HomeDir" => "0.91",
  "HTTP::Date" => 0,
  "Module::Build" => "0.28",
  "Test::More" => "0.94",
  "URI::file" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = My::ModuleBuild->new(%module_build_args);

$build->create_build_script;


lib/App/cpangitify.pm
package App::cpangitify;

use strict;
use warnings;
use autodie qw( :system );
use v5.10;
use Getopt::Long qw( GetOptions );
use Pod::Usage qw( pod2usage );
use Path::Class qw( file dir );
use Git::Wrapper;
use File::Temp qw( tempdir );
use File::chdir;
use JSON::PP qw( decode_json );
use URI;
use PerlX::Maybe qw( maybe );
use File::Copy::Recursive qw( rcopy );
use File::Basename qw( basename );
use Archive::Extract;
use File::Spec;
use CPAN::ReleaseHistory;
use HTTP::Tiny;

# ABSTRACT: Convert cpan distribution from BackPAN to a git repository
our $VERSION = '0.11'; # VERSION


our $ua  = HTTP::Tiny->new;
our $opt_metacpan_url;

sub _rm_rf
{
  my($file) = @_;
  
  if($file->is_dir && ! -l $file)
  {
    _rm_rf($_) for $file->children;
  }
  
  $file->remove || die "unable to delete $file";
}

sub main
{
  my $class = shift;
  local @ARGV = @_;
  
  my $opt_backpan_index_url;
  my $opt_backpan_url = "http://backpan.perl.org/authors/id";
  $opt_metacpan_url   = "http://api.metacpan.org/";

  GetOptions(
    'backpan_index_url=s' => \$opt_backpan_index_url,
    'backpan_url=s'       => \$opt_backpan_url,
    'metacpan_url=s'      => \$opt_metacpan_url,
    'help|h'              => sub { pod2usage({ -verbose => 2}) },
    'version'             => sub {
      say 'cpangitify version ', ($App::cpangitify::VERSION // 'dev');
      exit 1;
    },
  ) || pod2usage(1);

  my @names = map { s/::/-/g; $_ } @ARGV;
  my %names = map { $_ => 1 } @names;
  my $name = $names[0];

  pod2usage(1) unless $name;

  my $dest = dir()->absolute->subdir($name);

  if(-e $dest)
  {
    say "already exists: $dest";
    return 2;
  }

  say "creating/updating index...";
  my $history = CPAN::ReleaseHistory->new(
    maybe url => $opt_backpan_index_url
  )->release_iterator;

  say "searching...";
  my @rel;
  while(my $release = $history->next_release)
  {
    next unless $names{$release->distinfo->dist};
    push @rel, $release;
  }

  if($@ || @rel == 0)
  {
    say "no releases found for $name";
    return 2;
  }

  say "mkdir $dest";
  $dest->mkpath(0,0700);

  my $git = Git::Wrapper->new($dest->stringify);

  $git->init;

  sub author($)
  {
    state $cache = {};
  
    my $cpanid = shift;
  
    unless(defined $cache->{$cpanid})
    {
      my $uri = URI->new($opt_metacpan_url . "v0/author/" . $cpanid);
      my $res = $ua->get($uri);
      unless($res->{success})
      {
        say "error fetching $uri";
        say $res->{reason};
        return 2;
      }
      $cache->{$cpanid} = decode_json($res->{content})
    }
  
    sprintf "%s <%s>", $cache->{$cpanid}->{name}, $cache->{$cpanid}->{email}->[0];
  }

  foreach my $rel (@rel)
  {
    my $path    = $rel->path;
    my $version = $rel->distinfo->version;
    my $time    = $rel->timestamp;
    my $cpanid  = $rel->distinfo->cpanid;
  
    say "$path [ $version ]";
  
    my $tmp = dir( tempdir( CLEANUP => 1 ) );
  
    local $CWD = $tmp->stringify;
  
    my $uri = URI->new(join('/', $opt_backpan_url, $path));
    say "fetch ... $uri";
    my $res = $ua->get($uri);
    unless($res->{success})
    {
      say "error fetching $uri";
      say $res->{reason};
      return 2;
    }
  
    do {
      my $fn = basename $uri->path;
    
      open my $fh, '>', $fn;
      binmode $fh;
      print $fh $res->{content};
      close $fh;

      say "unpack... $fn";
      my $archive = Archive::Extract->new( archive => $fn );
      $archive->extract( to => File::Spec->curdir ) || die $archive->error;
      unlink $fn;
  
    };
  
    my $source = do {
      my @children = map { $_->absolute } dir()->children;
      if(@children != 1)
      {
        say "archive doesn't contain exactly one child: @children";
      }
  
      $CWD = $children[0]->stringify;
      $children[0];
    };
  
    say "merge...";
  
    foreach my $child ($dest->children)
    {
      next if $child->basename eq '.git';
      _rm_rf($child);
    }
  
    foreach my $child ($source->children)
    {
      if(-d  $child)
      {
        rcopy($child, $dest->subdir($child->basename)) || die "unable to copy $child $!";
      }
      else
      {
        rcopy($child, $dest->file($child->basename)) || die "unable to copy $child $!";
      }
    }
  
    say "commit and tag...";
    $git->rm($_->from) for grep { $_->mode eq 'deleted' } $git->status->get('changed');
    $git->add('.');
    $git->commit({
      message => "version $version",
      date    => "$time +0000",
      author  => author $cpanid,
    });
    eval { $git->tag($version) };
    warn $@ if $@;
  }
  
  return 0;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

App::cpangitify - Convert cpan distribution from BackPAN to a git repository

=head1 VERSION

version 0.11

=head1 DESCRIPTION

This is the module for the L<cpangitify> script.  See L<cpangitify> for details.

=head1 SEE ALSO

L<cpangitify>

=head1 AUTHOR

Graham Ollis <plicease@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;

use 5.010000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Convert cpan distribution from BackPAN to a git repository",
  "AUTHOR" => "Graham Ollis <plicease\@cpan.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Git::Wrapper" => 0,
    "Sort::Versions" => 0
  },
  "DISTNAME" => "App-cpangitify",
  "EXE_FILES" => [
    "bin/cpangitify"
  ],
  "LICENSE" => "perl",
  "NAME" => "App::cpangitify",
  "PREREQ_PM" => {
    "BackPAN::Index" => "0.41",
    "File::Copy::Recursive" => 0,
    "File::chdir" => 0,
    "Git::Wrapper" => 0,
    "IPC::System::Simple" => 0,
    "JSON" => 0,
    "LWP::UserAgent" => 0,
    "Path::Class" => 0,
    "PerlX::Maybe" => 0,
    "URI" => 0,
    "autodie" => 0
  },
  "TEST_REQUIRES" => {
    "Capture::Tiny" => 0,
    "File::HomeDir" => 0,
    "Test::More" => 0,
    "URI::file" => 0
  },
  "VERSION" => "0.07",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };


use Git::Wrapper;
use Sort::Versions qw( versioncmp );

if ( versioncmp( Git::Wrapper->new(".")->version , '1.5.0') eq -1 ) {
  print "git version 1.5.0 or better is required\n";
  exit 2;
}

WriteMakefile(%WriteMakefileArgs);



